### ISSUE TITLE
Disallow relative file includes

### ISSUE THREAD

**Issue by kkmuffme** [WordPress-Coding-Standards/WordPress-Coding-Standards#1630]
Like in VIP ruleset (which was removed from Wordpress ruleset recently and is now standalone):

https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/#relative-file-includes

**Comment by dingo-d** [comment]
Same as in the previously opened issue, VIP standard has been removed, as noted in the [Changelog.md](https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/blob/develop/CHANGELOG.md#200-rc1---2018-12-31).

Don't know if we have any rule about relative paths tbh.

**Comment by kkmuffme** [comment]
Yes I know, that's why I propose to add a rule for this in the WordPress core standards to check for this, as this makes sense for every WP plugin/code, not just for VIP code.

**Comment by lkraav** [comment]
> Including files with relative paths may lead to unintended results.

It's quite unfortunate VIP documentation doesn't give a single example of what an "unintended result" might be.

**Comment by kkmuffme** [comment]
Bc the "unintended" result is really dependent on the specific case as there are multiple possibilities I guess.
We also had the issue some time ago that relative paths were not working due to some specific set up constraints of our php install.

For example php ini allows setting an include_path, which would break use of relative file paths, or someone calls chdir to change the working directory, ...

**Comment by jrfnl** [comment]
While I agree that using absolute paths is preferred over relative paths, this is not something which is reliably sniffable.

It is a very common pattern for plugins/themes to define a constant for the path to the plugin/theme in the initial file and to use that throughout the rest of the plugin/theme. A sniff would not be aware of that, so would throw false positives 95% of the time, which is not acceptable.

Suggest: close.

**Comment by GaryJones** [comment]
The closest VIP currently has is [here](https://github.com/Automattic/VIP-Coding-Standards/blob/master/WordPressVIPMinimum/Sniffs/Files/IncludingFileSniff.php).

**Comment by kkmuffme** [comment]
It is easily sniffable by just checking that the path either contains " ." (in case a constant is used) or a / (when a path is used)

It's very simplified, but would like your feedback (just comment the various lines to test it)

```
// CORRECT:
$string = "include '/var/example.php';";
$string = "include( '/var/example.php' );";
$string = "include_once '/var/example.php';";
$string = "include MY_CONSTANT . 'example.php';";
$string = "include MY_CONSTANT.'example.php';";

// WRONG
$string = "include 'example.php';";

if ( preg_match( '/\.[ ]*\'|\.[ ]*"|\//', $string ) !== 1 ) {
    echo "wrong";
}
```

**Comment by jrfnl** [comment]
@kkmuffme I appreciate that you mean well, but what you are saying shows a lack of understanding of how code can be written and how sniffing code works.

Simple example of code which would break on your suggestion:
```php
// Example 1:
$filename = ABS_PATH . '/path/to/file'; // Note: this concatenation can be done in a completely different file.
include $filename;

// Example 2:
require_once RELATIVE_PATH . '/' . $file_name;
```

Asides from that, PHPCS works with the Tokenizer extension and using a regex to determine concatenation won't work and is the wrong way to go about things.

I've told you when you opened [an issue about this in the PHPCS repo](https://github.com/squizlabs/PHP_CodeSniffer/issues/2374) to move it to the Automattic VIP repo. Instead you moved it to WPCS.
I'm telling you again now: VIP rules should be addressed in the [Automattic VIP repo](https://github.com/Automattic/VIP-Coding-Standards/).

As I said before, while the principle of this rule is valid, this is not reliably sniffable and therefore not suitable for WPCS.
